test/utils/bytecode.hpp

```diff
@@ -90,30 +91,79 @@ big_endian(T value)
     return {static_cast<uint8_t>(value >> 8), static_cast<uint8_t>(value)};
 }
 
-inline bytecode eof_header(
-    uint8_t version, uint16_t code_size, uint16_t max_stack_height, uint16_t data_size)
+struct eof_bytecode
 {
-    bytecode out{bytes{0xEF, 0x00, version}};
-    out += "01" + big_endian(uint16_t{4});  // type header
-    out += "02"_hex + big_endian(uint16_t{1}) + big_endian(code_size);
-    out += "04" + big_endian(data_size);
-    out += "00";
-    out += "0080"_hex + big_endian(max_stack_height);  // type section
-    return out;
-}
+private:
+    std::vector<bytecode> m_codes;
+    std::vector<evmone::EOFCodeType> m_types;
+    bytecode m_data;
 
-inline bytecode eof1_header(uint16_t code_size, uint16_t max_stack_height, uint16_t data_size = 0)
-{
-    return eof_header(1, code_size, max_stack_height, data_size);
-}
+    bytecode eof_header() const
+    {
+        assert(!m_codes.empty());
+        assert(m_codes[0].size() <= std::numeric_limits<uint16_t>::max());
+        assert(m_data.size() <= std::numeric_limits<uint16_t>::max());
+        assert(m_codes.size() == m_types.size());
+
+        constexpr uint8_t version = 0x01;
+
+        bytecode out{bytes{0xEF, 0x00, version}};
+
+        // type header
+        const auto types_size = static_cast<uint16_t>(m_types.size() * 4);
+        out += "01" + big_endian(types_size);
+
+        // codes header
+        const auto code_count = static_cast<uint16_t>(m_codes.size());
+        out += "02"_hex + big_endian(code_count);
+        for (const auto& code : m_codes)
+        {
+            const auto code_size = static_cast<uint16_t>(code.size());
+            out += big_endian(code_size);
+        }
+
+        // data header
+        const auto data_size = static_cast<uint16_t>(m_data.size());
+        out += "04" + big_endian(data_size);
+        out += "00";  // terminator
+
+        // types section
+        for (const auto& type : m_types)
+            out += bytes{type.inputs, type.outputs} + big_endian(type.max_stack_height);
+        return out;
+    }
+
+public:
+    explicit eof_bytecode(bytecode code, uint16_t max_stack_height = 0)
+      : m_codes{std::move(code)}, m_types{{0, 0x80, max_stack_height}}
+    {}
+
+    auto& code(bytecode c, uint8_t inputs, uint8_t outputs, uint8_t max_stack_height)
+    {
+        m_codes.emplace_back(std::move(c));
+        m_types.emplace_back(inputs, outputs, max_stack_height);
+        return *this;
+    }
+
+    auto& data(bytecode d)
+    {
+        m_data = std::move(d);
+        return *this;
+    }
+
+    operator bytecode() const
```

Comment:
```suggestion
    explicit(false) operator bytecode() const
```
Welcome to modern C++.

Reply:
This is the same as default without it?